File: backend/nodejs/functions/routes/studyPlan.js
const express = require("express");
const router = express.Router();
const admin = require("firebase-admin");
const db = admin.firestore();

// plan kaydetme
router.post("/", async (req, res) => {
  try {
    const {userId, plans, meta} = req.body;

    const userRef = db.collection("study-plan").doc(userId);
    await userRef.set({
      weeklyTotalHours: meta.weeklyTotalHours,
      aiConfidence: meta.aiConfidence,
    });

    const batch = db.batch();
    plans.forEach((plan) => {
      const planRef = userRef.collection("plans").doc(plan.date);

      batch.set(planRef, {
        day: plan.day,
        dailyTotalHours: plan.dailyTotalHours,
        plan: plan.sessions.map((s) => ({
          id: s.id,
          startingTime: s.startingTime,
          endTime: s.endTime,
          lessons: s.lessons,
          lessonsSubject: s.lessonsSubject,
          lessonsDuration: s.lessonsDuration,
          completed: false,
        })),
      });
    });

    await batch.commit();
    res.status(200).json({message: "Plans + meta saved successfully"});
  } catch (err) {
    console.error("POST / hata:", err);
    res.status(500).json({error: err.message});
  }
});

// tüm planları getir
router.get("/:userId", async (req, res) => {
  try {
    const {userId} = req.params;
    const userRef = db.collection("study-plan").doc(userId);
    const userDoc = await userRef.get();
    const meta = userDoc.exists ? userDoc.data() : null;
    const snapshot = await userRef.collection("plans").get();
    const plans = snapshot.docs.map((doc) => ({
      date: doc.id,
      ...doc.data(),
    }));

    res.json({
      meta: meta || {
        weeklyTotalHours: 0.0,
        aiConfidence: 0.0,
      },
      plans: plans,
    });
  } catch (err) {
    console.error("GET /:userId hata:", err);
    res.status(500).json({error: err.message});
  }
});

// id ile plan getirme
router.get("/:userId/:planDate", async (req, res) => {
  try {
    const {userId, planDate} = req.params;
    const planRef = db
        .collection("study-plan")
        .doc(userId)
        .collection("plans")
        .doc(planDate);

    const doc = await planRef.get();
    if (!doc.exists) return res.status(404).json({error: "Plan not found"});
    res.json({date: planDate, ...doc.data()});
  } catch (err) {
    console.error("GET /:userId/:planDate hata:", err);
    res.status(500).json({error: err.message});
  }
});

// ders tamamlama durumu
router.put("/:userId/:planDate/session/:sessionId", async (req, res) => {
  try {
    const {userId, planDate, sessionId} = req.params;
    const {completed} = req.body;

    console.log("PUT /session toggle çağrıldı");
    console.log({userId, planDate, sessionId, completed});

    const planRef = db
        .collection("study-plan")
        .doc(userId)
        .collection("plans")
        .doc(planDate);

    const planDoc = await planRef.get();
    if (!planDoc.exists) {
      console.log("Plan bulunamadı");
      return res.status(404).json({error: "Plan not found"});
    }

    const planData = planDoc.data();
    if (!planData || !planData.plan) {
      console.log("Plan içinde session yok");
      return res.status(404).json({error: "Sessions not found in plan"});
    }
    const updatedSessions = planData.plan.map((session) => {
      if (session.id === sessionId) {
        console.log(`Session bulundu, completed güncelleniyor: ${completed}`);
        return {...session, completed: completed};
      }
      return session;
    });
    await planRef.set({plan: updatedSessions}, {merge: true});

    res.status(200).json({message: "Session completion status updated"});
  } catch (err) {
    console.error("PUT /session toggle hata:", err);
    res.status(500).json({error: err.message});
  }
});

module.exports = router;

File: frontend/vena/lib/features/calendar/data/model/study_plan.dart
import 'study_session.dart';

class StudyPlan {
  final String date;
  final String day;
  final List<StudySession> sessions;
  final double dailyTotalHours;
  double weeklyTotalHours;
  double aiConfidence;

  StudyPlan({
    required this.date,
    required this.day,
    required this.sessions,
    required this.dailyTotalHours,
    this.weeklyTotalHours = 0.0,
    this.aiConfidence = 0.0,
  });

  factory StudyPlan.fromJson(String date, Map<String, dynamic> json) {
    final day = json['day'] ?? '';
    final plan = (json['plan'] as List<dynamic>?) ?? [];

    final sessions = plan
        .map((sessionJson) => StudySession.fromJson(sessionJson))
        .toList();

    return StudyPlan(
      date: date,
      day: day,
      sessions: sessions,
      dailyTotalHours: (json['dailyTotalHours'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'day': day,
      'plan': sessions.map((s) => s.toJson()).toList(),
      'dailyTotalHours': dailyTotalHours,
    };
  }

  StudyPlan copyWith({
    String? date,
    String? day,
    List<StudySession>? sessions,
    double? dailyTotalHours,
    double? weeklyTotalHours,
    double? aiConfidence,
    
  }) {
    return StudyPlan(
      date: date ?? this.date,
      day: day ?? this.day,
      sessions: sessions ?? this.sessions,
      dailyTotalHours: dailyTotalHours ?? this.dailyTotalHours,
      weeklyTotalHours: weeklyTotalHours ?? this.weeklyTotalHours,
      aiConfidence: aiConfidence ?? this.aiConfidence,
    );
  }
}
